### ISSUE TITLE
etcd 3.5.0 resurrects ancient (unremovable) members

### ISSUE THREAD

**Issue by anguslees** [etcd-io/etcd#13196]
Zombie members!

I [upgraded](https://github.com/anguslees/k8s-home/commit/b883c6d761f0a8a882a0b5ff1a5b46b46c5a0e36) my home k8s cluster to etcd 3.5.0 (from 3.4.14).  I noticed in the logs that it started complaining about the health of surprising IPs on my network, but it didn't seem to affect other functionality.  Error messages like:
```
{"level":"warn","ts":"2021-06-16T08:40:21.800Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_SNAPSHOT","remote-peer-id":"8e9e05c52164694d","rtt":"0s","error":"dial tcp 192.168.0.9:2380: connect: no route to host"}
```
Where `192.168.0.9:2380` is _not_ one of my current etcd cluster members.  It's an ancient one from a _long_ time ago.  I did some git archaeology, and I think this member last existed back when I was using etcd 3.3.15 (or even earlier).

Now I note `etcdctl member list` includes all these ancient members, which it did not before the upgrade:
```console
roachbait:~% kubectl -n kube-system exec -ti etcd-2 -- etcdctl member list
407604e04d511f63, started, kmaster2, https://192.168.0.140:2380, https://192.168.0.140:2379, false
80df0211ad8a11c3, started, 0b5642a6cc18493d81a606483d9cbb7b, https://192.168.0.132:2380, https://192.168.0.132:2379, false
8e9e05c52164694d, started, default, https://192.168.0.9:2380, http://127.0.0.1:2379,https://192.168.0.9:2379, false
aaaacf0529a32f2f, started, default, https://192.168.0.9:2380, https://192.168.0.9:2379, false
b5ed3c441abd0677, unstarted, , https://192.168.0.40:2380, , false
c7a95206b6a22d21, started, fc4698cdc1184810a2c3447a7ee66689, https://192.168.0.129:2380, https://192.168.0.129:2379, false
c93ba62f815b42ed, started, 765885d83e774555bc7ee0f9c6fc1178, https://192.168.0.117:2380, https://192.168.0.117:2379, true
cc28792212a199be, started, 887f1b514ea54520a61643163d427d42, https://192.168.0.161:2380, https://192.168.0.161:2379, false
```
(407604e04d511f63 8e9e05c52164694d aaaacf0529a32f2f b5ed3c441abd0677 were removed a long time ago and should not be in this list.)

... and I can't seem to remove these zombie members:

```console
roachbait:~% kubectl -n kube-system exec -ti etcd-2 -- etcdctl member remove 407604e04d511f63
{"level":"warn","ts":"2021-06-16T08:42:46.434Z","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0002f08c0/#initially=[https://192.168.0.129:2379/]","attempt":0,"error":"rpc error: code = NotFound desc = etcdserver: member not found"}
Error: etcdserver: member not found
command terminated with exit code 1
```

My member list _should_ be only 192.168.0.{129, 132, 161, 117}.  Note that 192.168.0.117 is a 'learner' (hot spare) which might be a bit unusual.

Otherwise, it appears the cluster is working fine, so I presume it still thinks the quorum is only 2-out-of-3:
```console
roachbait:~% kubectl -n kube-system exec -ti etcd-2 -- etcdctl endpoint health
https://192.168.0.129:2379/ is healthy: successfully committed proposal: took = 27.37035ms
```

```console
roachbait:~% kubectl -n kube-system exec -ti etcd-2 -- etcdctl version                       
etcdctl version: 3.5.0
API version: 3.5
```

I note 3.5.0 release notes mention a change to the way membership is reported.  I'm guessing that's related somehow.

**Comment by brenix** [comment]
We have run across a similar and what seems to be the same issue across several clusters that have been upgraded over time (<3.4.x). Clusters recently provisioned with 3.4.x do not seem to exhibit the same issue when upgraded.

Upon upgrading, we see a duplicate of each member (with a different id). The cluster continues to operate, however the logs are filled with warnings.

Attempts to remove the zombie members also shows "etcdserver: member not found".

Only workaround that seems to work is to start with a fresh cluster and restore from snapshot, which isn't ideal when availability is a concern. We have since rolled back to 3.4.x series to avoid this until a solution can be found.

**Comment by allenporter** [comment]
Similar, after upgrade to 3.5 things look good:
```
$ etcdctl endpoint status
cfg01.prod:2379, b9428e2a6571f403, 3.5.0, 103 MB, false, 8048, 65770996
cfg02.prod:2379, 71d808e4df973c61, 3.5.0, 106 MB, true, 8048, 65770996
cfg03.prod:2379, 24cfcc8c8928a23, 3.5.0, 105 MB, false, 8048, 65770997
```

However, seeing many old members rather than just the expected 3:
```
$ etcdctl member list
24cfcc8c8928a23, started, cfg03, http://10.10.21.26:2380, http://10.10.21.26:2379,http://127.0.1.1:2379
57179c3602d3017, started, cfg03, http://10.10.21.26:2380, http://10.10.21.26:2379
71d808e4df973c61, started, cfg02, http://10.10.21.25:2380, http://10.10.21.25:2379,http://127.0.1.1:2379
774d4084c61cdba9, started, cfg02, http://10.10.21.25:2380, http://10.10.21.25:2379
a53e9eafb0791018, started, cfg01, http://10.10.21.27:2380, http://10.10.21.27:2379,http://127.0.1.1:2379
b05ffc3c3e406282, started, cfg01, http://10.10.21.27:2380, http://10.10.21.27:2379
b9428e2a6571f403, started, cfg01, http://10.10.21.27:2380, http://10.10.21.27:2379,http://127.0.1.1:2379
cfaa8b1d45d54d63, started, cfg03, http://10.10.21.26:2380, http://10.10.21.26:2379
e2e14bd3143026eb, started, cfg03, http://10.10.21.26:2380, http://10.10.21.26:2379,http://127.0.1.1:2379
f41b8cf40578fb5b, started, cfg02, http://10.10.21.25:2380, http://10.10.21.25:2379
fb45dc04aab58183, started, cfg01, http://10.10.21.27:2380, http://10.10.21.27:2379
```

**Comment by ptabor** [comment]
Could you, please, check/provide what following commands return on your DB files ?
*members* and *members_removed" sections are especially important

```
% gobin -run go.etcd.io/bbolt/cmd/bbolt keys tmp/snap/db members
61a3717f95707065

% gobin -run go.etcd.io/bbolt/cmd/bbolt get tmp/snap/db members 61a3717f95707065
{"id":703213213213352,"peerURLs":["http://foo-bar:2380"],"name":"foo-bar-vm","clientURLs":["http://127.0.0.1:2379"]}

% gobin -run go.etcd.io/bbolt/cmd/bbolt keys tmp/snap/db members_removed                

% cat tmp/snap/0000000000000008-0000000001a0c25b.snap
 // after copy-pasting the json part, and e.g. formatting using https://jsonformatter.curiousconcept.com/#
and checking there is no sensitive content)

e.g. :

   "Root":{
      "Path":"/",
      "CreatedIndex":0,
      "ModifiedIndex":0,
      "ExpireTime":"0001-01-01T00:00:00Z",
      "Value":""
      "Children":{
         "0":{
            "Path":"/0",
            "CreatedIndex":0,
            "ModifiedIndex":0,
            "ExpireTime":"0001-01-01T00:00:00Z",
            "Value":"",
            "Children":{
               "members":{
                  "Path":"/0/members",
                  "CreatedIndex":1,
                  "ModifiedIndex":1,
                  "ExpireTime":"0001-01-01T00:00:00Z",
                  "Value":"",
                  "Children":{
                     "61a3717f95707065":{
                        "Path":"/0/members/61a3717f95707065",
                        "CreatedIndex":1,
                        "ModifiedIndex":1,
                        "ExpireTime":"0001-01-01T00:00:00Z",
                        "Value":"",
                        "Children":{
                           "attributes":{
                              "Path":"/0/members/61a3717f95707065/attributes",
                              "CreatedIndex":9,
                              "ModifiedIndex":9,
                              "ExpireTime":"0001-01-01T00:00:00Z",
                              "Value":"{\"name\":\"foo-bar\",\"clientURLs\":[\"http://127.0.0.1:2379\"]}",
                              "Children":null
                           },
                           "raftAttributes":{
                              "Path":"/0/members/61a3717f95707065/raftAttributes",
                              "CreatedIndex":1,
                              "ModifiedIndex":1,
                              "ExpireTime":"0001-01-01T00:00:00Z",
                              "Value":"{\"peerURLs\":[\"http://foo-vm:2380\"]}",
                              "Children":null
                           }
                        }
                     }
                  }
               },
               "version":{
                  "Path":"/0/version",
                  "CreatedIndex":3,
                  "ModifiedIndex":3,
                  "ExpireTime":"0001-01-01T00:00:00Z",
                  "Value":"3.2.0",
                  "Children":null
               }
            }
         },
         "1":{
            "Path":"/1",
            "CreatedIndex":0,
            "ModifiedIndex":0,
            "ExpireTime":"0001-01-01T00:00:00Z",
            "Value":"",
            "Children":{
               
            }
         }
      }
   },
   "WatcherHub":{
      "EventHistory":{
         "Queue":{
            "Events":[
               {
                  "action":"create",
                  "node":{
                     "key":"/0/members/61a3717f95707065/raftAttributes",
                     "value":"{\"peerURLs\":[\"foo-bar\"]}",
                     "modifiedIndex":1,
                     "createdIndex":1
                  }
               },
               {
                  "action":"set",
                  "node":{
                     "key":"/0/members/61a3717f95707065/attributes",
                     "value":"{\"name\":\"foo-bar\",\"clientURLs\":[\"http://127.0.0.1:2379\"]}",
                     "modifiedIndex":2,
                     "createdIndex":2
                  }
               },
               {
                  "action":"set",
                  "node":{
                     "key":"/0/version",
                     "value":"3.2.0",
                     "modifiedIndex":3,
                     "createdIndex":3
                  }
               },
               {
                  "action":"set",
                  "node":{
                     "key":"/0/members/61a3717f95707065/attributes",
                     "value":"{\"name\":\"foo-vm-3f76-710f-vm\",\"clientURLs\":[\"http://127.0.0.1:2379\"]}",
                     "modifiedIndex":4,
                     "createdIndex":4
                  },
                  "prevNode":{
                     "key":"/0/members/61a3717f95707065/attributes",
                     "value":"{\"name\":\"etcd-foo-vm\",\"clientURLs\":[\"http://127.0.0.1:2379\"]}",
                     "modifiedIndex":2,
                     "createdIndex":2
                  }
               },
               {
                  "action":"set",
                  "node":{
                     "key":"/0/members/61a3717f95707065/attributes",
                     "value":"{\"name\":\"foo-vm-3f76-8ca9-vm\",\"clientURLs\":[\"http://127.0.0.1:2379\"]}",
                     "modifiedIndex":5,
                     "createdIndex":5
                  },
                  "prevNode":{
                     "key":"/0/members/61a3717f95707065/attributes",
                     "value":"{\"name\":\"foo-vm-3f76-710f-vm\",\"clientURLs\":[\"http://127.0.0.1:2379\"]}",
                     "modifiedIndex":4,
                     "createdIndex":4
                  }
               },
               {
                  "action":"set",
                  "node":{
                     "key":"/0/members/61a3717f95707065/attributes",
                     "value":"{\"name\":\"foo-vm-3f76-0dc4-vm\",\"clientURLs\":[\"http://127.0.0.1:2379\"]}",
                     "modifiedIndex":6,
                     "createdIndex":6
                  },
                  "prevNode":{
                     "key":"/0/members/61a3717f95707065/attributes",
                     "value":"{\"name\":\"foo-vm-3f76-8ca9-vm\",\"clientURLs\":[\"http://127.0.0.1:2379\"]}",
                     "modifiedIndex":5,
                     "createdIndex":5
                  }
               },
               {
                  "action":"set",
                  "node":{
                     "key":"/0/members/61a3717f95707065/attributes",
                     "value":"{\"name\":\"foo-vm-3f76-da7e-vm\",\"clientURLs\":[\"http://127.0.0.1:2379\"]}",
                     "modifiedIndex":7,
                     "createdIndex":7
                  },
                  "prevNode":{
                     "key":"/0/members/61a3717f95707065/attributes",
                     "value":"{\"name\":\"foo-vm-3f76-0dc4-vm\",\"clientURLs\":[\"http://127.0.0.1:2379\"]}",
                     "modifiedIndex":6,
                     "createdIndex":6
                  }
               },
               {
                  "action":"set",
                  "node":{
                     "key":"/0/members/61a3717f95707065/attributes",
                     "value":"{\"name\":\"foo-vm-3f76-19b2-vm\",\"clientURLs\":[\"http://127.0.0.1:2379\"]}",
                     "modifiedIndex":8,
                     "createdIndex":8
                  },
                  "prevNode":{
                     "key":"/0/members/61a3717f95707065/attributes",
                     "value":"{\"name\":\"foo-vm-3f76-da7e-vm\",\"clientURLs\":[\"http://127.0.0.1:2379\"]}",
                     "modifiedIndex":7,
                     "createdIndex":7
                  }
               },
               {
                  "action":"set",
                  "node":{
                     "key":"/0/members/61a3717f95707065/attributes",
                     "value":"{\"name\":\"foo-vm-3f76-4333-vm\",\"clientURLs\":[\"http://127.0.0.1:2379\"]}",
                     "modifiedIndex":9,
                     "createdIndex":9
                  },
                  "prevNode":{
                     "key":"/0/members/61a3717f95707065/attributes",
                     "value":"{\"name\":\"foo-vm-3f76-19b2-vm\",\"clientURLs\":[\"http://127.0.0.1:2379\"]}",
                     "modifiedIndex":8,
                     "createdIndex":8
                  }
               },}
   ...
}
```

**Comment by ptabor** [comment]
I tried to repro this (without success) by: 
  1. Starting new etcd-3.3.25 cluster
  2. Storing some content. 
  3. Expanding it to make it 2 member cluster etcdctl v2
  4. Populating content to trigger a snapshot
  5. Removing the member using etcdctl v3 
  6.  Populating content to trigger a new snapshot
  7. Running etcd 3.5 
  8. Verifying etcdctl (v3) output. There is only one member as expected.

**Comment by allenporter** [comment]
Thank you. I am unable to reproduce and get data now as my cluster failed and I had to restore from a snapshot. (The root cause of that failure was not related to this issue)

In my case step #5 involved many of the other runtime reconfiguration steps like replacing failed members or adding and removing members (with a previously used name) over the course of a year.

**Comment by anguslees** [comment]
> Could you, please, check/provide what following commands return on your DB files ?

`bbolt keys tmp/snap/db members`

```
407604e04d511f63
80df0211ad8a11c3
8e9e05c52164694d
aaaacf0529a32f2f
b5ed3c441abd0677
c7a95206b6a22d21
c93ba62f815b42ed
cc28792212a199be
```

`for k in $(bbolt keys tmp/snap/db members); do bbolt get tmp/snap/db members $k; done`

```
{"id":4644905427095592803,"peerURLs":["https://192.168.0.140:2380"],"name":"kmaster2","clientURLs":["https://192.168.0.140:2379"]}
{"id":9286143231610458563,"peerURLs":["https://192.168.0.132:2380"],"name":"0b5642a6cc18493d81a606483d9cbb7b","clientURLs":["https://192.168.0.132:2379"]}
{"id":10276657743932975437,"peerURLs":["https://192.168.0.9:2380"],"name":"default","clientURLs":["http://127.0.0.1:2379","https://192.168.0.9:2379"]}
{"id":12297869353568907055,"peerURLs":["https://192.168.0.9:2380"],"name":"default","clientURLs":["https://192.168.0.9:2379"]}
{"id":13109200353549420151,"peerURLs":["https://192.168.0.40:2380"]}
{"id":14387120673399123233,"peerURLs":["https://192.168.0.129:2380"],"name":"fc4698cdc1184810a2c3447a7ee66689","clientURLs":["https://192.168.0.129:2379"]}
{"id":14500366148213359341,"peerURLs":["https://192.168.0.117:2380"],"isLearner":true,"name":"765885d83e774555bc7ee0f9c6fc1178","clientURLs":["https://192.168.0.117:2379"]}
{"id":14711141370054154686,"peerURLs":["https://192.168.0.161:2380"],"name":"887f1b514ea54520a61643163d427d42","clientURLs":["https://192.168.0.161:2379"]}
```

`bbolt keys tmp/snap/db members_removed`

```
291aee36546cf7d9
35e881d8acafab93
543eb708915700fa
5e96932294f1e6fc
63cc74379832b19e
6e2d84e43ce7f5dc
886b9831c2bc6344
8c57f5f478292992
8ca0979a7be988fb
912b5ec6a4351e72
a9d03a65b24424cf
b0d64bf3a7dc029e
b7f74bdc6f870e55
c6c25a20c2ab96cf
d9071c73d3580ee9
e279f72b5d4c0938
e4aa84d5f44a3595
f20ca578eb614628
```

`tmp/snap/0000000000114d8c-0000000013464bce.snap`

->  https://gist.github.com/anguslees/7335f8088bb85b0cd2aee708d4982545

**Comment by smira** [comment]
might be related to #13119 which is also about having zombie members

**Comment by ptabor** [comment]
I think that the problem has origins semewhere in v3.1-v3.3, i.e. this versions had bugs that lead to store v2 get out of sync with store v3. Till etcd v.3.4, storeV2 was considered authoritative. v3.5 started to consider store-v3 as authoritative. 

I think we need to add logic at the start of etcd that if both available (v2 & v3), makes sure v3 is in sync with v2. 
If v2 is not available, it would just depend on v3 (so allow potential rollback from v3.6 where store v2 snapshots might not be populated).

**Comment by anguslees** [comment]
Whatever the source of truth is, `member remove` should operate on that, and not something else 😛

**Comment by ptabor** [comment]
It operates on both and fails if any of them failed: 
  https://github.com/etcd-io/etcd/blob/946a5a6f25c3b6b89408ab447852731bde6e6289/server/etcdserver/api/membership/cluster.go#L403-L412

**Comment by serathius** [comment]
I can take a look into implementing this. My current plan is to implement syncing bbolt to v2store state if it's available in current master and cherry pick it to v3.5. As result we fix will be available in v3.6 and v3.5.1.

Fixing it just in v3.5 is not enough as this will leave upgrade path v3.5.0 -> v3.6.0 broken. One consequence of this would be that we will need to postpone removal of v2store totally to v3.7. In v3.6 the bbolt becomes source of truth, but only after we make sure it's synced up with v2store. Please let me know if that makes sense.

**Comment by ptabor** [comment]
Thank you @serathius ! 

Postponing cleanup of store-v2 code by one release (so another year or so) is a taught choice, but formally correct. I find it unlikely that someone in practice:
  1. got store-v2 diverged from store-v3 in etcd v3.2... 
  2. upgraded to v3.5.0 and not noticed the problem. Didn't upgraded to v3.5.1+ that have 'syncing code'. 
  3. upgraded to v3.6.x (with store-v2 gone completely) and noticed the problem that late.  

So would be leaning towards following AIs: 
   1.  Make sure RemoveMember/AddMember is idempotent, i.e. work in v3.5 on both data-sources independently and succeeds as long as the intended state is met (given member is or is-not in the final state of both storages). 
 Thanks to this even if user will discover the problem in v3.6, they have a way to fix it. 

  2. Make in v3.5.1+ only code that reads both store-v2 (only if exists) and store-v3 and syncs v2->v3 if they diverge. 

I hope that both changes combined would still allow deprecating store-v2 in etcd-v3.6.

**Comment by serathius** [comment]
I think I found an problem with implementing sync between v2->v3. From what I understand v2 storage is not saved imminently to disk which means that when reading members (during server bootstrap in Recover function) state of v2 and v3 is expected to diverge. 

I discovered this via TestV3HashRestart failing, in this test server is restarted to compare v3 database hash before and after restart. I found that hash changed as V3 store members were overwritten by V2 store members. V2 store members were empty as Recover happends before reading WAL entries.

**Comment by zapotah** [comment]
I can confirm that after upgrade to 3.5.1 via k8s 1.23.0, this issue has been solved at least for my test cases where this manifested.

**Comment by peterska** [comment]
Can this backported to 1.22.X please. My cluster will not upgrade from 1.21.7 to 1.22.4 due to this issue. It gets stuck waiting for etcd to be ready. Checking kube-system/etcd logs shows it is trying to contact the zombie etcd node. This is a single node etcd kubernetes cluster created using kubeadm a long time ago. I migrated the cluster so that the api sever is accessed using a dns name rather than an ip address and have since changed the ip address . This what caused the phantom etcd member.

**Comment by ptabor** [comment]
k8s backport strategy is being discussed in: https://github.com/kubernetes/kubernetes/issues/106589. Closing.

**Comment by serathius** [comment]
@jiapeish, issue should be fixed in v3.5.1 allowing you to remove ancient members. Please ensure you run fixed version, executed proper steps to mitigate the issue. To followup on your comment and reopen the issue we need more precise information about steps you made.

**Comment by jiapeish** [comment]
> @jiapeish, issue was resolved in in v3.5.1, by allowing you to remove ancient members. Ancient members are still expected to appear if state of members in your cluster was corrupted (V2 and V3 storage desynced). However additional members should not have any impact on cluster health or quorum. Please upgrade your cluster to v3.5.1 and remove them manually.
> 
> To followup on your comment and reopen the issue we need more precise information about steps you made to confirm that fix is not working for you.

@serathius The details is decribed here [13697](https://github.com/etcd-io/etcd/issues/13697)

**Comment by ahrtr** [comment]
similar issue https://github.com/etcd-io/etcd/issues/20967
